posts/convenient-secrets-with-spring-boot-cloud-aws.md

```diff
@@ -0,0 +1,106 @@
+- In the *text* field, enter `ABC123`.  
+- Click `Create Parameter` to save it.
+
+A few things to note, for the *type* field we could of used a String but instead I chose SecureString to keep the value encrypted at rest.  You can do either and it will be transparently decrypted when it is retrieved.  For the `name` field, the prefix `/config/application/` is important but everything else is just made up.  I'll go into why the prefix is important in just a bit.
+
+
+## Retrieving Secrets
+
+The "secret" sauce of this solution is in the integration.  Now that your configuration is securely stored in AWS, you need a good way to retrieve it.  A poor way to integrate would be to specifiy each property in the ECS task definition using ValueFrom mappings.  That's just more configuration and more to maintain.  Unsurprisingly, there's a great Spring integration that can help us out called **Spring Cloud Starter AWS Parameter Store Config**.  
+
+### Step 1 - Add the dependency
+
+In your application's pom.xml, add:
+
+```xml
+<dependency>
+  <groupId>org.springframework.cloud</groupId>
+  <artifactId>spring-cloud-starter-aws-parameter-store-config</artifactId>
+</dependency>
+```
+
+Additionally, if Spring Cloud isn't setup in the project yet, add this to the pom.xml as well:
+
+```xml
+<project>
+   ...
+
+  <properties>
+    ...
+    <!-- This is the latest version as of the 7/7/2020 -->
+    <spring-cloud.version>Hoxton.SR6</spring-cloud.version> 
+  </properties>
+   
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>org.springframework.cloud</groupId>
+        <artifactId>spring-cloud-dependencies</artifactId>
+        <version>${spring-cloud.version}</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+</project>
+```
+
+
+### Step 2 - Deploy!
+
+That is it.  And now, there is some magic happening as is the case with most Spring integrations.
+
+Deploy your application to AWS and Spring Cloud will access the Parameter Store on AWS.  Properties are retrieved and injected based on the following **prefix** conventions:
+
+- `/config/application/` - applies to all applications 
+- `/config/application_dev/` - applies to all applications with an active `dev` profile 
+- `/config/my-api/` -  applies to only the `my-api` application (defined by `spring.application.name`)
+- `/config/my-api_dev` -  applies to only the `my-api` application with an active `dev` profile  
+
+These conventions provide the flexibility to define global, application-specific, and environment-specific parameters.  An added benefit is that parameters can be added/changed/removed in AWS Parameter Store and will sync on application restart.  There's no middle layers or task definitions to update.
+
+# Wrapping Up
+
+If this seems simple, it's because it is.  Spring Boot's starter packages tend to do a good job of the box with minimal to no configuration.  

```

Comment:
Spring Boot's starter packages tend to do a good job `out-of-the-box` with `little` to no configuration.